design/resource.go

```diff
@@ -0,0 +1,104 @@
+package design
+
+import (
+	d "github.com/goadesign/goa/design"
+	a "github.com/goadesign/goa/design/apidsl"
+)
+
+var _ = a.Resource("resource", func() {
+
+	a.BasePath("/resource")
+
+	a.DefaultMedia(ResourceMedia)
+
+	a.Action("register", func() {
+		a.Routing(
+			a.POST(""),
+		)
+		a.Description("Register a new resource")
+		a.Payload(ResourceMedia)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.Created, JSONAPIErrors)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+	})
+
+	a.Action("read", func() {
+		a.Routing(
+			a.GET("/:resourceId"),
+		)
+		a.Params(func() {
+			a.Param("resourceId", d.String, "The identifier of the resource to read")
+		})
+		a.Description("Read a specific resource")
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.TemporaryRedirect)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("update", func() {
+		a.Routing(
+			a.PUT("/:resourceId"),
+		)
+		a.Params(func() {
+			a.Param("resourceId", d.String, "Identifier of the resource to update")
+		})
+		a.Description("Update the details of the specified resource")
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.TemporaryRedirect)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("delete", func() {
+		a.Routing(
+			a.DELETE("/:resourceId"),
+		)
+		a.Params(func() {
+			a.Param("resourceId", d.String, "Identifier of the resource to delete")
+		})
+		a.Description("Delete a resource")
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.TemporaryRedirect)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("list", func() {
+		a.Routing(
+			a.GET(""),
+		)
+		a.Description("List all resources")
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.TemporaryRedirect)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+	})
+})
+
+// ResourceMedia represents a protected resource
+var ResourceMedia = a.MediaType("application/vnd.resource+json", func() {
+	a.Description("A Protected Resource")
+	a.Attributes(func() {
+		a.Attribute("resource_scopes", a.ArrayOf(d.String), "The valid scopes for this resource")
+		a.Attribute("icon_uri", d.String, "The URI containing the icon for the resource")
```

Comment:
Do we really need any icons here? I know that UMA specify them. But shouldn't we just ignore them?

Reply:
My opinion is that we allow the parameter in order to be compliant with UMA (while compliance might not be important right now it might be in the future) however we just don't do anything with it if one is passed in a request. So yes, we just ignore it.

Reply:
Sure. I'm not saying we should forbid this param. But we don't have to declare this param to allow it.  It will be allowed anyway. But by declaring it we generate additional code (this param will be included in goa model, context, tests, etc).

Reply:
So a JSON document passed in the request body does not have to conform with the MediaType definition declared for that method?  Won't there be some kind of validation error if the document contains properties that the MediaType does not declare?

Reply:
Oh, didn't notice it's a media type. It's not requests parameters. But anyway additional (not declared) attributes are just ignored. So, it's safe not to declare them. Payloads with undeclared attributes are still valid.